File: mock.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"os"
	"strings"
)

type HandlerFunc func(http.ResponseWriter, *http.Request)

type Response struct {
	Status int
	Body   string
}

type Params map[string]string

func (p Params) Matches(query url.Values) bool {
	for k, v := range p {
		if query.Get(k) != v {
			return false
		}
	}

	for k := range query {
		if query.Get(k) != p[k] {
			return false
		}
	}

	return true
}

type Request struct {
	Url      string
	Method   string
	Params   Params
	Data     string
	Response Response
	Called   bool
}

func (r *Request) Matches(req *http.Request) bool {
	return req.Method == r.Method && r.Params.Matches(req.URL.Query())
}

type MocksVerification struct {
	AllCalled bool `json:"allCalled"`
}

type Mocks map[string]*Request

func (m *Mocks) AddFromRequest(req *http.Request) error {
	dec := json.NewDecoder(req.Body)
	var r Request
	if err := dec.Decode(&r); err == nil || err == io.EOF {
		(*m)[r.Url] = &r
	} else {
		return err
	}
	return nil
}

func (m *Mocks) Reset() error {
	for k := range *m {
		delete(*m, k)
	}
	return nil
}

func (m *Mocks) ResponseFor(r *http.Request) (*Request, Response, bool) {
	req, found := (*m)[r.URL.Path]
	if found {
		if req.Matches(r) {
			return req, req.Response, true
		} else {
			log.Printf("Hit %s but didn't match. Expected params: %s. Given params: %s.\n", r.URL.Path, req.Params, r.URL.Query())
			return nil, Response{}, false
		}
	}
	return nil, Response{}, false
}

func (m *Mocks) Verification() (MocksVerification, error) {
	for _, r := range *m {
		if !r.Called {
			return MocksVerification{AllCalled: false}, nil
		}
	}
	return MocksVerification{AllCalled: true}, nil
}

func exists(path string) (bool, error) {
	_, err := os.Stat(path)

	if err == nil {
		return true, nil
	}

	if os.IsNotExist(err) {
		return false, nil
	}

	return true, err
}

func mocksHandler(mocks Mocks) HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		switch r.Method {
		case "GET":
			if verification, err := mocks.Verification(); err != nil {
				log.Println(err)
			} else {
				enc := json.NewEncoder(w)
				enc.Encode(verification)
			}
		case "POST":
			if err := mocks.AddFromRequest(r); err != nil {
				log.Println(err)
			}
		case "DELETE":
			if err := mocks.Reset(); err != nil {
				log.Println(err)
			}
		}
	}
}

func mocksReporter(mocks Mocks) HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		io.WriteString(w, fmt.Sprintf("%s", mocks))
	}
}

func staticFile(w http.ResponseWriter, path string) {
	var contentType string
	if strings.HasSuffix(path, ".css") {
		contentType = "text/css"
	}
	w.Header().Add("Content-Type", contentType)

	f, err := os.Open(path)
	if err != nil {
		log.Println(err)
	} else {
		io.Copy(w, f)
	}
}

func any(mocks Mocks, staticRoot string) HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var staticPath string
		if r.URL.Path == "/" {
			staticPath = staticRoot + "/index.html"
		} else {
			staticPath = staticRoot + r.URL.Path
		}

		if found, err := exists(staticPath); err != nil {
			log.Println(err)
			return
		} else if found {
			staticFile(w, staticPath)
		} else if request, response, found := mocks.ResponseFor(r); found {
			request.Called = true
			if response.Status != 0 {
				w.WriteHeader(response.Status)
			} else {
				w.WriteHeader(http.StatusOK)
			}
			io.WriteString(w, response.Body)
		} else {
			w.WriteHeader(http.StatusNotFound)
		}
	}
}

func main() {
	port := flag.String("port", "8080", "TCP port to bind to")
	staticRoot := flag.String("root", ".", "Root directory")

	flag.Parse()

	mocks := Mocks{}

	http.HandleFunc("/mocks", mocksHandler(mocks))
	http.HandleFunc("/meta", mocksReporter(mocks))
	http.HandleFunc("/", any(mocks, *staticRoot))

	log.Println("Serving on port", *port)
	log.Fatal(http.ListenAndServe(":"+*port, nil))
}

File: mock.js
var http = require('http'),
  path = require('path'),
  fs = require('fs'),
  url = require('url'),
  querystring = require('querystring');

var mocks = {};

function parsePath(request) {
  return url.parse(request.url).pathname;
}

function allCalled() {
  for(var p in mocks) {
    if(mocks.hasOwnProperty(p) && !mocks[p].called) {
      return false;
    }
  }
  return true;
}

function verify(response) {
  response.writeHead(200, {'Content-Type': 'application/json'});
  response.write(JSON.stringify({allCalled: allCalled()}));
  response.end();
}

function addMock(request, response) {
  var body = '';

  request.on('data', function(chunk) {
    body += chunk.toString();
  });

  request.on('end', function() {
    var m = JSON.parse(body);
    mocks[m.url] = m;

    response.writeHead(204);
    response.end();
  });
}

function handleMock(request, response) {
  switch (request.method) {
    case 'GET':
      verify(response);
      break;
    case 'POST':
      addMock(request, response);
      break;
    case 'DELETE':
      mocks = {};
      response.writeHead(204);
      response.end();
      break;
  }
}

function matchesParams(expected, actual) {
  for(var k in expected) {
    if(expected.hasOwnProperty(k) && expected[k] !== actual[k]) {
      return false;
    }
  }

  for(var k in actual) {
    if(actual.hasOwnProperty(k) && expected[k] !== actual[k]) {
      return false;
    }
  }

  return true;
}

function matches(expected, actual) {
  var basic = expected && expected.method === actual.method;
  if(basic) {
    if(matchesParams(expected.params, querystring.parse(url.parse(actual.url).query))) {
      return true;
    } else {
      console.warn('Hit '+expected.url+' but did not match. Expected params '+JSON.stringify(expected.params)+'. '+
                   'Given params '+JSON.stringify(querystring.parse(url.parse(actual.url).query))+'.');
      return false;
    }
  } else {
    return false;
  }
}

function isMocked(request) {
  var mock = mocks[parsePath(request)];
  return mock && matches(mock, request);
}

function returnMocked(request, response) {
  var mock = mocks[parsePath(request)];
  mock.called = true;
  response.writeHeader(mock.status || 200, {'Content-Type': 'application/json'});
  response.write(mock.response.body);
  response.end();
}

function returnStatic(filename, response) {
  if(fs.statSync(filename).isDirectory()) {
    filename += '/index.html';
  }

  fs.readFile(filename, 'binary', function(err, file) {
    if(err) {
      response.writeHead(500, {'Content-Type': 'application/json'});
      response.write(JSON.stringify({error: err}));
      response.end();
    } else {
      var contentType = '';
      if (/\.css$/.test(filename)) {
        contentType = 'text/css';
      }

      response.writeHead(200, {'Content-Type': contentType});
      response.write(file, 'binary');
      response.end();
    }
  });
}

function handleRequest(staticRoot, request, response) {
  var filename = path.join(staticRoot, parsePath(request));
  fs.exists(filename, function(exists) {
    if(exists){
      returnStatic(filename, response);
    } else if (isMocked(request)) {
      returnMocked(request, response);
    } else {
      response.writeHead(404, {'Content-Type': 'application/json'});
      response.write('{}');
      response.end();
    }
  });
}

exports.startServer = function(port, staticRoot) {
  var server = http.createServer(function(request, response) {
    var p = parsePath(request);
    if (/^\/mocks$/.test(p)) {
      handleMock(request, response);
    } else if (/^\/meta$/.test(p)) {
      response.writeHead(200, {'Content-Type': 'application/json'});
      response.write(JSON.stringify(mocks));
      response.end();
    } else {
      handleRequest(staticRoot, request, response);
    }
  }).listen(port);
  console.log('Mock API server running on port ' + port);
  return server;
}
